### PR TITLE
IDEMPIERE-5238 - Support of Search display type for AD_Table_ID

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA Ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5238
Comment: https://idempiere.atlassian.net/browse/IDEMPIERE-5238?focusedCommentId=49239

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have made corresponding changes to the documentation as follows:
 - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
   - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
   - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by hengsin** [comment]
hi, both of your new PR mess up the PR check list.

**Comment by PeterTakacs300** [comment]
oh ok thanks, fixed

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
Testing this threw NPE when selecting RV_OpenItem in the table info window:

```
15:34:40.373===========> UiEngineImpl.error:  [213]
java.lang.NullPointerException
	at org.adempiere.webui.editor.WRecordIDEditor.getRecordsLookup(WRecordIDEditor.java:373)
	at org.adempiere.webui.window.WRecordIDDialog.valueChange(WRecordIDDialog.java:225)
	at org.adempiere.webui.editor.WEditor.fireValueChange(WEditor.java:472)
	at org.adempiere.webui.editor.WSearchEditor.actionCombo(WSearchEditor.java:543)
	at org.adempiere.webui.editor.WSearchEditor$2.onEvent(WSearchEditor.java:709)
	at org.zkoss.zk.ui.AbstractComponent.onEvent(AbstractComponent.java:3191)
	at org.zkoss.zk.ui.AbstractComponent.service(AbstractComponent.java:3161)
	at org.zkoss.zk.ui.AbstractComponent.service(AbstractComponent.java:3103)
	at org.zkoss.zk.ui.impl.EventProcessor.process(EventProcessor.java:138)
	at org.zkoss.zk.ui.event.Events.sendEvent(Events.java:367)
	at org.adempiere.webui.component.Window.onPageDetached(Window.java:89)
	at org.adempiere.webui.panel.InfoPanel.onPageDetached(InfoPanel.java:3034)
	at org.zkoss.zk.ui.AbstractComponent.setPage0(AbstractComponent.java:520)
	at org.zkoss.zk.ui.AbstractComponent.setPageBefore(AbstractComponent.java:434)
	at org.zkoss.zk.ui.AbstractComponent.setPage(AbstractComponent.java:397)
	at org.zkoss.zk.ui.AbstractComponent.detach(AbstractComponent.java:1968)
	at org.adempiere.webui.panel.InfoPanel.lambda$1(InfoPanel.java:2888)
	at org.zkoss.zk.ui.impl.DesktopImpl$ScheduleInfo.invoke(DesktopImpl.java:1856)
	at org.zkoss.zk.ui.impl.DesktopImpl$ScheduleInfo.access$1100(DesktopImpl.java:1843)
	at org.zkoss.zk.ui.impl.DesktopImpl$ScheduleListener.onEvent(DesktopImpl.java:1867)
	at org.zkoss.zk.ui.AbstractComponent.onEvent(AbstractComponent.java:3191)
```

**Comment by PeterTakacs300** [comment]
Hi @CarlosRuiz-globalqss

The problem was with tables that has no Key Column.

 I've added a "This Table has no Key Column" error message/exception to this use case. @norbertbede suggests to add a filter to the Table field when used in Record ID editor, which would exclude these tables from the list by default. wdyt?

**Comment by CarlosRuiz-globalqss** [comment]
Please move scripts to iD10

**Comment by PeterTakacs300** [comment]
Done

**Comment by hengsin** [comment]
[pr1726.search.patch.txt](https://github.com/idempiere/idempiere/files/10970261/pr1726.search.patch.txt)
- alternative patch to use TableDir when AD_Table_ID display type is Search.

Ignore this if your prefer the original approach in this PR.

**Comment by PeterTakacs300** [comment]
@CarlosRuiz-globalqss - the bug is fixed

@hengsin - thanks for the patch, @norbertbede had a similar idea to force Table Dir editor, the question is can it be a performance issue if we don't have any dynamic validation (e.g. the user wants to have all the tables there - even if this use case is unlikely) on the AD_Table_ID field and then it will load all the tables into the Table Dir editor.

**Comment by hengsin** [comment]
ok, I guess it make sense to have choice here. perhaps we should have a build in info window for table in future, the general one is just not good.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WRecordIDDialog.java [hunk 1]
```diff
@@ -117,12 +124,22 @@ private void init(Page page) {
...
+    	}
+    	tableIDEditor.addValueChangeListener(this);
 		tableIDEditor.setValue(tableIDValue);
+		
 		int tableID = tableIDValue != null ? tableIDValue.intValue() : 0;
-		recordsEditor = tableID > 0 ? new WSearchEditor("Record_ID", false, false, true, editor.getRecordsLookup(tableID)) : null;
-    	
+		MLookup recordsLookup = editor.getRecordsLookup(tableID);
+		if(recordsLookup != null)
+			recordsEditor = tableID > 0 ? new WSearchEditor("Record_ID", false, false, true, recordsLookup) : null;
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WRecordIDDialog.java:141]: you are doing duplicate checking here, if recordsLookup is not null, tableID must > 0.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WRecordIDDialog.java [hunk 2]
```diff
@@ -189,31 +205,40 @@ public void onEvent(Event event) throws Exception {
...
+			if (recordsEditor != null) {
+				recordsEditor.setValue(null);
+				recordsEditorLabel.detach();
+				recordsEditor.getComponent().detach();
+			}
+			int tableID = Integer.parseInt(Objects.toString(evt.getNewValue(), "-1"));
+			if(tableID > 0) {
+				MLookup recordsLookup = editor.getRecordsLookup(tableID);
+				if(recordsLookup != null) {
+					recordsEditor = tableID > 0 ? new WSearchEditor("Record_ID", false, false, true, recordsLookup) : null;
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WRecordIDDialog.java:232]: again, duplicate checking for tableID > 0

Comment (PeterTakacs300) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WRecordIDDialog.java]: multiple tableID > 0 checks fixed

**Review by CarlosRuiz-globalqss — APPROVED** [review]

**Comment by CarlosRuiz-globalqss** [comment]
Hi @hengsin - my latest tests passed - I didn't do code peer review.

**Review by hengsin — APPROVED** [review]
looks good and tested by Carlos

**Comment by CarlosRuiz-globalqss** [comment]
> I've added a "This Table has no Key Column" error message/exception to this use case. @norbertbede suggests to add a filter to the Table field when used in Record ID editor, which would exclude these tables from the list by default. wdyt?

Hi @PeterTakacs300 / @norbertbede 

There is a reference with name "AD_Table with ID Column" - maybe this reference can do the trick to filter the tables you want here?
